<a href="https://colab.research.google.com/github/venkataraghuramr/Hamoeye-Data-Science-Internship/blob/main/StageA_Introduction_to_Python_for_Machine_Learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
df = pd.read_csv("FoodBalanceSheets_E_Africa_NOFLAG.csv",encoding = 'latin-1')

In [ ]:
df

Area Code      Area  Item Code           Item  Element Code  \
0              4   Algeria       2501     Population           511   
1              4   Algeria       2501     Population          5301   
2              4   Algeria       2901    Grand Total           664   
3              4   Algeria       2901    Grand Total           674   
4              4   Algeria       2901    Grand Total           684   
...          ...       ...        ...            ...           ...   
60938        181  Zimbabwe       2899  Miscellaneous          5142   
60939        181  Zimbabwe       2899  Miscellaneous           645   
60940        181  Zimbabwe       2899  Miscellaneous           664   
60941        181  Zimbabwe       2899  Miscellaneous           674   
60942        181  Zimbabwe       2899  Miscellaneous           684   

                                      Element             Unit     Y2014  \
0               Total Population - Both sexes     1000 persons  38924.00   
1                    Domestic supply quantity      1000 tonnes      0.00   
2               Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3      Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4          Fat supply quantity (g/capita/day)     g/capita/day     80.06   
...                                       ...              ...       ...   
60938                                    Food      1000 tonnes     42.00   
60939     Food supply quantity (kg/capita/yr)               kg      3.06   
60940           Food supply (kcal/capita/day)  kcal/capita/day      3.00   
60941  Protein supply quantity (g/capita/day)     g/capita/day      0.10   
60942      Fat supply quantity (g/capita/day)     g/capita/day      0.04   

          Y2015     Y2016     Y2017     Y2018  
0      39728.00  40551.00  41389.00  42228.00  
1          0.00      0.00      0.00      0.00  
2       3379.00   3372.00   3341.00   3322.00  
3         94.35     94.72     92.82     91.83  
4         79.36     77.40     80.19     77.28  
...         ...       ...       ...       ...  
60938     46.00     33.00     19.00     16.00  
60939      3.33      2.35      1.33      1.08  
60940      4.00      3.00      1.00      1.00  
60941      0.11      0.08      0.04      0.04  
60942      0.05      0.03      0.02      0.01  

[60943 rows x 12 columns]

In [ ]:
df.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

In [ ]:
df.dtypes

Area Code         int64
Area             object
Item Code         int64
Item             object
Element Code      int64
Element          object
Unit             object
Y2014           float64
Y2015           float64
Y2016           float64
Y2017           float64
Y2018           float64
dtype: object

In [ ]:
df.isnull().sum()

Area Code          0
Area               0
Item Code          0
Item               0
Element Code       0
Element            0
Unit               0
Y2014           1589
Y2015           1548
Y2016           1535
Y2017           1506
Y2018           1436
dtype: int64

## Question 2 What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

In [ ]:
df.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

## Question 3 What is the total number and percentage of missing data in 2014 to 3 decimal places?
Ans - 1589

In [ ]:
df.isnull().sum()

Area Code          0
Area               0
Item Code          0
Item               0
Element Code       0
Element            0
Unit               0
Y2014           1589
Y2015           1548
Y2016           1535
Y2017           1506
Y2018           1436
dtype: int64

## Question 4 What is the Protein Supply in Madagascar in 2015

Ans - 173.05

In [ ]:
df.groupby('Element')['Area','Element','Y2015'].sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


Y2015
Element                                           
Domestic supply quantity                2021493.55
Export Quantity                          157614.47
Fat supply quantity (g/capita/day)        10235.74
Feed                                     225050.22
Food                                    1232361.10
Food supply (kcal/capita/day)            453383.00
Food supply quantity (kg/capita/yr)       49345.13
Import Quantity                          267018.46
Losses                                   155439.00
Other uses (non-food)                     66254.41
Processing                               287929.00
Production                              1947019.39
Protein supply quantity (g/capita/day)    11833.95
Residuals                                 30045.00
Seed                                      23976.82
Stock Variation                           34910.99
Total Population - Both sexes           1058081.00
Tourist consumption                         349.00

In [ ]:
x = df.groupby(['Element','Area'],sort = False)['Area','Element','Y2015'].sum()
x

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


Y2015
Element                                Area              
Total Population - Both sexes          Algeria   39728.00
Domestic supply quantity               Algeria   85972.86
Food supply (kcal/capita/day)          Algeria   13512.00
Protein supply quantity (g/capita/day) Algeria     377.37
Fat supply quantity (g/capita/day)     Algeria     317.40
...                                                   ...
Processing                             Zimbabwe   7914.00
Residuals                              Zimbabwe   -780.00
Food                                   Zimbabwe   8674.69
Food supply quantity (kg/capita/yr)    Zimbabwe    627.92
Other uses (non-food)                  Zimbabwe    173.13

[803 rows x 1 columns]

In [ ]:
x.to_csv('ra.csv') ### as it is becoming more tough tought of just saving a new csv file ang getting the value from there

## Question 10 Perform a groupby operation on element. What year has the highest sum of Stock Variation

Ans - Y2017

In [ ]:
df.groupby(['Element'],sort = False)['Y2015','Y2016','Y2017','Y2018'].sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


Y2015       Y2016       Y2017  \
Element                                                                      
Total Population - Both sexes           1058081.00  1085107.00  1112641.00   
Domestic supply quantity                2021493.55  2044842.70  2088198.10   
Food supply (kcal/capita/day)            453383.00   451810.00   454681.00   
Protein supply quantity (g/capita/day)    11833.95    11779.69    11842.45   
Fat supply quantity (g/capita/day)        10235.74    10102.77    10253.84   
Production                              1947019.39  1943537.15  2030056.89   
Import Quantity                          267018.46   286582.78   294559.09   
Stock Variation                           34910.99    33140.12    54316.91   
Export Quantity                          157614.47   151920.46   182338.80   
Feed                                     225050.22   228958.65   223705.68   
Seed                                      23976.82    23389.20    24870.14   
Losses                                   155439.00   157787.00   160614.00   
Processing                               287929.00   280631.00   292836.00   
Other uses (non-food)                     66254.41    69563.68    91645.97   
Residuals                                 30045.00    37224.00    35500.00   
Food                                    1232361.10  1247022.17  1258888.28   
Food supply quantity (kg/capita/yr)       49345.13    48985.28    48690.04   
Tourist consumption                         349.00       89.00       91.00   

                                             Y2018  
Element                                             
Total Population - Both sexes           1140605.00  
Domestic supply quantity                2161192.10  
Food supply (kcal/capita/day)            455261.00  
Protein supply quantity (g/capita/day)    11833.56  
Fat supply quantity (g/capita/day)        10258.69  
Production                              2075072.89  
Import Quantity                          287997.09  
Stock Variation                           20577.91  
Export Quantity                          181594.80  
Feed                                     233489.68  
Seed                                      25263.14  
Losses                                   163902.00  
Processing                               308429.00  
Other uses (non-food)                     91300.97  
Residuals                                 34864.00  
Food                                    1303841.28  
Food supply quantity (kg/capita/yr)       49056.85  
Tourist consumption                          90.00

## Question 11 What is the total number of unique countries in the dataset

Ans - 49

In [ ]:
question11 = df['Area'].unique()

In [ ]:
len(question11)

49

## Question 12 Select columns 'Y2017' AND 'Area'. Perform a group operation on 'Area'. Which of these areas had highest sum in 2017.

Ans - Nigeria

In [ ]:
question12 = df.groupby('Area')['Area','Y2017'].sum()
sortedDF = question12.sort_values('Y2017', ascending=False)
print(sortedDF)

                                  Y2017
Area                                   
Nigeria                      1483268.23
Egypt                         866379.92
South Africa                  517590.54
Ethiopia                      448683.76
Morocco                       388495.36
Ghana                         337599.06
Algeria                       325644.27
United Republic of Tanzania   322616.85
Kenya                         264660.66
Sudan                         239931.92
Cameroon                      232030.43
Angola                        229159.57
Côte d'Ivoire                 224599.01
Uganda                        213950.38
Malawi                        181098.71
Mozambique                    161407.98
Mauritania                    156665.46
Mali                          149928.33
Madagascar                    131197.73
Niger                         126707.58
Benin                         124771.22
Tunisia                       124167.20
Zambia                        103223.77


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


## Question 15 Select columns 'Y2017' AND 'Area'. Which of these Areas had the 7th lowest sum in 2017.

Ans - Guinea-Bissau

In [ ]:
question12 = df.groupby('Area')['Area','Y2017'].sum()
sortedDF = question12.sort_values('Y2017', ascending=True)
print(sortedDF)

                                  Y2017
Area                                   
Sudan (former)                     0.00
Ethiopia PDR                       0.00
Comoros                           59.84
Seychelles                       442.34
Sao Tome and Principe          12662.63
Cabo Verde                     14650.74
Guinea-Bissau                  19102.77
Lesotho                        21267.96
Botswana                       22101.30
Djibouti                       22729.91
Gambia                         23154.18
Gabon                          27979.64
Liberia                        29342.20
Namibia                        29874.89
Central African Republic       29937.00
Congo                          41181.68
Togo                           49841.88
Mauritius                      51114.83
Eswatini                       54343.33
Sierra Leone                   55311.33
Chad                           71594.68
Rwanda                         73663.69
Zimbabwe                       75919.34


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


## Question 17 What is the total sum of Wine produced in 2015 and 2018 respectively.

Ans 4251.81	4039.32

In [ ]:
df.groupby('Item')['Item','Y2015','Y2018'].sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


Y2015      Y2018
Item                                     
Alcohol, Non-Food      2180.00    2293.00
Alcoholic Beverages   98783.72   97847.27
Animal Products       11811.73   11578.61
Animal fats          200675.72  269648.27
Apples and products   10559.15    9640.51
...                        ...        ...
Vegetables, Other    158104.08  163987.21
Vegetal Products     107064.17  107775.39
Wheat and products   234710.51  242645.19
Wine                   4251.81    4039.32
Yams                 203151.78  221272.09

[119 rows x 2 columns]

## Question 19 Perform a groupby operation on Element. What is the total number of the sum of Processing in 2017

Ans 292836.00

In [ ]:
df.groupby('Element')['Element','Y2017'].sum() 

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


Y2017
Element                                           
Domestic supply quantity                2088198.10
Export Quantity                          182338.80
Fat supply quantity (g/capita/day)        10253.84
Feed                                     223705.68
Food                                    1258888.28
Food supply (kcal/capita/day)            454681.00
Food supply quantity (kg/capita/yr)       48690.04
Import Quantity                          294559.09
Losses                                   160614.00
Other uses (non-food)                     91645.97
Processing                               292836.00
Production                              2030056.89
Protein supply quantity (g/capita/day)    11842.45
Residuals                                 35500.00
Seed                                      24870.14
Stock Variation                           54316.91
Total Population - Both sexes           1112641.00
Tourist consumption                          91.00

## Question 20  Which year had the least coorelation with Element Code

Ans- 2016

In [ ]:
df.corr()

Area Code  Item Code  Element Code     Y2014     Y2015  \
Area Code      1.000000  -0.005159     -0.000209  0.006164  0.005472   
Item Code     -0.005159   1.000000     -0.024683  0.021722  0.020857   
Element Code  -0.000209  -0.024683      1.000000  0.024457  0.023889   
Y2014          0.006164   0.021722      0.024457  1.000000  0.994647   
Y2015          0.005472   0.020857      0.023889  0.994647  1.000000   
Y2016          0.005247   0.020109      0.023444  0.996081  0.995739   
Y2017          0.005006   0.021494      0.024254  0.995230  0.988048   
Y2018          0.005665   0.021314      0.024279  0.994872  0.988208   

                 Y2016     Y2017     Y2018  
Area Code     0.005247  0.005006  0.005665  
Item Code     0.020109  0.021494  0.021314  
Element Code  0.023444  0.024254  0.024279  
Y2014         0.996081  0.995230  0.994872  
Y2015         0.995739  0.988048  0.988208  
Y2016         1.000000  0.992785  0.992757  
Y2017         0.992785  1.000000  0.998103  
Y2018         0.992757  0.998103  1.000000